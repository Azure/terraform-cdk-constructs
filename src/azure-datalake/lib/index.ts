export * from "./filesystem";
export * from "./path";
export * from "./datalake";
