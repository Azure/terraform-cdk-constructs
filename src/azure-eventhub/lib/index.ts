export * from "./authorization";
export * from "./eventhubcluster";
export * from "./consumer";
export * from "./instance";
export * from "./kusto-connection";
export * from "./namespace";
