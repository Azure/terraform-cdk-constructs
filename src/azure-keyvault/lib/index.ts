export * from './certificate'
export * from './key'
export * from './policy'
export * from './secret'
export * from './vault'

