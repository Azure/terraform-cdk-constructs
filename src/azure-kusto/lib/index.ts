export * from "./cluster";
export * from "./compute-specification";
export * from "./database";
