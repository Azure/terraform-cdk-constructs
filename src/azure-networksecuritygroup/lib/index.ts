export * from './preconfigured-rules';
export * from './securitygroup';
