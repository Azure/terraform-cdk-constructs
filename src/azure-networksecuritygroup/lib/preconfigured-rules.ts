import { RuleConfig } from ".";

/**
 * Properties for defining overrides for a rule in an Azure Network Security Group.
 */
export interface RuleOverrides {
  /**
   * Optional source address prefix to be matched for the rule. This can be an IP address or a range of IP addresses.
   * If not specified, the default behavior is to match any source address.
   */
  readonly sourceAddressPrefix?: string;

  /**
   * Optional destination address prefix to be matched for the rule. Similar to the source address prefix,
   * this can be a specific IP address or a range. If not provided, it defaults to matching any destination address.
   */
  readonly destinationAddressPrefix?: string;

  /**
   * Optional priority for the rule. Rules are processed in the order of their priority,
   * with lower numbers processed before higher numbers. If not provided, a default priority will be assigned.
   */
  readonly priority?: number;
}

export class PreconfiguredRules {
  // Active Directory
  static activeDirectoryAllowADReplication: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "*",
    sourcePortRange: "*",
    destinationPortRange: "389",
    name: "AllowADReplication",
    priority: 601,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  static activeDirectoryAllowADReplicationSSL: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "*",
    sourcePortRange: "*",
    destinationPortRange: "636",
    name: "AllowADReplicationSSL",
    priority: 602,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  static activeDirectoryAllowADGCReplication: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "*",
    sourcePortRange: "*",
    destinationPortRange: "3268",
    name: "AllowADGCReplication",
    priority: 603,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  static activeDirectoryAllowADGCReplicationSSL: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "*",
    sourcePortRange: "*",
    destinationPortRange: "3269",
    name: "AllowADGCReplicationSSL",
    priority: 604,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  static activeDirectoryAllowDNS: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "*",
    sourcePortRange: "*",
    destinationPortRange: "53",
    name: "AllowDNS",
    priority: 605,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  static activeDirectoryAllowKerberosAuthentication: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "*",
    sourcePortRange: "*",
    destinationPortRange: "88",
    name: "AllowKerberosAuthentication",
    priority: 606,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  static activeDirectoryAllowADReplicationTrust: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "*",
    sourcePortRange: "*",
    destinationPortRange: "445",
    name: "AllowADReplicationTrust",
    priority: 607,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  static activeDirectoryAllowSMTPReplication: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "25",
    name: "AllowSMTPReplication",
    priority: 608,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  static activeDirectoryAllowRPCReplication: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "135",
    name: "AllowRPCReplication",
    priority: 609,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  static activeDirectoryAllowFileReplication: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "5722",
    name: "AllowFileReplication",
    priority: 610,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  static activeDirectoryAllowWindowsTime: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "UDP",
    sourcePortRange: "*",
    destinationPortRange: "123",
    name: "AllowWindowsTime",
    priority: 611,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  static activeDirectoryAllowPasswordChangeKerberes: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "*",
    sourcePortRange: "*",
    destinationPortRange: "464",
    name: "AllowPasswordChangeKerberes",
    priority: 612,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  static activeDirectoryAllowDFSGroupPolicy: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "UDP",
    sourcePortRange: "*",
    destinationPortRange: "138",
    name: "AllowDFSGroupPolicy",
    priority: 613,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  static activeDirectoryAllowADDSWebServices: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "9389",
    name: "AllowADDSWebServices",
    priority: 614,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  static activeDirectoryAllowNETBIOSAuthentication: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "UDP",
    sourcePortRange: "*",
    destinationPortRange: "137",
    name: "AllowNETBIOSAuthentication",
    priority: 615,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  static activeDirectoryAllowNETBIOSReplication: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "139",
    name: "AllowNETBIOSReplication",
    priority: 616,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  // Cassandra
  static cassandra: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "9042",
    name: "Cassandra",
    priority: 551,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  static cassandraJmx: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "7199",
    name: "Cassandra-JMX",
    priority: 552,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  static cassandraThrift: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "9160",
    name: "Cassandra-Thrift",
    priority: 553,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  // CouchDB
  static couchDb: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "5984",
    name: "CouchDB",
    priority: 554,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  static couchDbHttps: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "6984",
    name: "CouchDB-HTTPS",
    priority: 555,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  // DNS-Tcp
  static dnsTcp: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "53",
    name: "DNS-Tcp",
    priority: 520,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  // DNS-UDP
  static dnsUdp: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "UDP",
    sourcePortRange: "*",
    destinationPortRange: "53",
    name: "DNS-UDP",
    priority: 521,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  // DynamicPorts
  static dynamicPorts: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "49152-65535",
    name: "DynamicPorts",
    priority: 650,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  // ElasticSearch
  static elasticSearch: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "9200-9300",
    name: "ElasticSearch",
    priority: 556,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  // FTP
  static ftp: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "21",
    name: "FTP",
    priority: 512,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  // HTTP-Tcp
  static httpTcp: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "80",
    name: "HTTP",
    priority: 513,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  // HTTP-UDP
  static httpUdp: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "UDP",
    sourcePortRange: "*",
    destinationPortRange: "80",
    name: "HTTP",
    priority: 514,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  // HTTPS-Tcp
  static https: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "443",
    name: "HTTPS",
    priority: 515,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  // IMAP
  static imap: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "143",
    name: "IMAP",
    priority: 516,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  // IMAPS
  static imaps: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "993",
    name: "IMAPS",
    priority: 516,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  // Kestrel
  static kestrel: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "22133",
    name: "Kestrel",
    priority: 517,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  // LDAP
  static ldap: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "389",
    name: "LDAP",
    priority: 558,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  // MongoDB
  static mongoDB: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "27017",
    name: "MongoDB",
    priority: 559,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  // Memcached
  static memcached: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "11211",
    name: "Memcached",
    priority: 560,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  // MSSQL
  static mssql: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "1433",
    name: "MSSQL",
    priority: 561,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  // MySQL
  static mySQL: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "3306",
    name: "MySQL",
    priority: 562,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  // Neo4J
  static neo4J: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "7474",
    name: "Neo4J",
    priority: 563,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  // POP3
  static pop3: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "110",
    name: "POP3",
    priority: 564,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  // POP3S
  static pop3s: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "995",
    name: "POP3S",
    priority: 565,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  // PostgreSQL
  static postgreSQL: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "5432",
    name: "PostgreSQL",
    priority: 566,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  // RabbitMQ
  static rabbitMQ: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "5672",
    name: "RabbitMQ",
    priority: 567,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  // RDP
  static rdp: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "3389",
    name: "RDP",
    priority: 505,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  // Redis
  static redis: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "6379",
    name: "Redis",
    priority: 568,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  // Riak
  static riak: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "8093",
    name: "Riak",
    priority: 568,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  // RiakJMX
  static riakJMX: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "8985",
    name: "Riak-JMX",
    priority: 569,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  // SMTP
  static smtp: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "25",
    name: "SMTP",
    priority: 570,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  // SMTPS
  static smtps: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "465",
    name: "SMTPS",
    priority: 571,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  // SSH
  static ssh: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "22",
    name: "SSH",
    priority: 501,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  // WinRM
  static winRM: RuleConfig = {
    direction: "Inbound",
    access: "Allow",
    protocol: "Tcp",
    sourcePortRange: "*",
    destinationPortRange: "5986",
    name: "WinRM",
    priority: 502,
    sourceAddressPrefix: "*",
    destinationAddressPrefix: "*",
  };

  static applyRuleOverrides(
    baseRule: RuleConfig,
    overrides: RuleOverrides,
  ): RuleConfig {
    return {
      ...baseRule,
      ...overrides,
    };
  }

  static addSourceAddress(
    rule: RuleConfig,
    sourceAddressPrefix: string,
  ): RuleConfig {
    return { ...rule, sourceAddressPrefix };
  }

  static addDestinationAddress(
    rule: RuleConfig,
    destinationAddressPrefix: string,
  ): RuleConfig {
    return { ...rule, destinationAddressPrefix };
  }

  static addPriority(rule: RuleConfig, priority: number): RuleConfig {
    return { ...rule, priority };
  }
}
