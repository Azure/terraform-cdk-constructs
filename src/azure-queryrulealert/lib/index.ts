export * from './query-rule-alert'