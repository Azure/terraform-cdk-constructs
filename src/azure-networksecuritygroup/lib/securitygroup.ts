import { NetworkInterface } from "@cdktf/provider-azurerm/lib/network-interface"; // Import the NetworkInterface class
import { NetworkInterfaceSecurityGroupAssociation } from "@cdktf/provider-azurerm/lib/network-interface-security-group-association";
import { NetworkSecurityGroup } from "@cdktf/provider-azurerm/lib/network-security-group";
import { NetworkSecurityRule } from "@cdktf/provider-azurerm/lib/network-security-rule";
import { ResourceGroup } from "@cdktf/provider-azurerm/lib/resource-group";
import { Subnet } from "@cdktf/provider-azurerm/lib/subnet";
import { SubnetNetworkSecurityGroupAssociation } from "@cdktf/provider-azurerm/lib/subnet-network-security-group-association";
import { Construct } from "constructs";
import { AzureResource } from "../../core-azure/lib";

/**
 * Configuration properties for defining a rule within an Azure Network Security Group.
 */
export interface RuleConfig {
  /**
   * The name of the security rule.
   */
  readonly name: string;

  /**
   * The priority of the rule. Lower numbers have higher priority. Allowed values are from 100 to 4096.
   */
  readonly priority: number;

  /**
   * The direction of the rule, which can be 'Inbound' or 'Outbound'.
   */
  readonly direction: string;

  /**
   * The access type of the rule, which determines whether the rule permits or denies traffic. Common values are 'Allow' or 'Deny'.
   */
  readonly access: string;

  /**
   * The protocol to which the rule applies, such as 'Tcp', 'Udp', or '*' (for all protocols).
   */
  readonly protocol: string;

  /**
   * The range of source ports to which the rule applies. Can be a single port or a range like '1024-2048'.
   */
  readonly sourcePortRange: string;

  /**
   * The range of destination ports to which the rule applies. Can also be a single port or a range.
   */
  readonly destinationPortRange: string;

  /**
   * The CIDR or source IP range or '*' to match any IP. This is the range of source IPs for which the rule applies.
   */
  readonly sourceAddressPrefix: string;

  /**
   * The CIDR or destination IP range or '*' to match any IP. This specifies the range of destination IPs for which the rule is applicable.
   */
  readonly destinationAddressPrefix: string;
}

/**
 * Properties for defining an Azure Network Security Group.
 */
export interface SecurityGroupProps {
  /**
   * The name of the resource group under which the network security group will be created.
   */
  readonly resourceGroup: ResourceGroup;

  /**
   * The Azure region in which to create the network security group, e.g., 'East US', 'West Europe'.
   */
  readonly location: string;

  /**
   * The name of the network security group. Must be unique within the resource group.
   */
  readonly name: string;

  /**
   * An array of rule configurations to be applied to the network security group.
   */
  readonly rules: RuleConfig[];
}

export class SecurityGroup extends AzureResource {
  readonly props: SecurityGroupProps;
  public id: string;
  public readonly name: string;
  public resourceGroup: ResourceGroup;

  constructor(scope: Construct, id: string, props: SecurityGroupProps) {
    super(scope, id);

    this.props = props;
    this.resourceGroup = props.resourceGroup;

    // Create a network security group
    const nsg = new NetworkSecurityGroup(this, "nsg", {
      name: props.name,
      resourceGroupName: props.resourceGroup.name,
      location: props.location,
    });

    // Create security rules within the network security group
    for (const ruleConfig of props.rules) {
      new NetworkSecurityRule(this, ruleConfig.name, {
        name: ruleConfig.name,
        resourceGroupName: props.resourceGroup.name,
        networkSecurityGroupName: nsg.name,
        priority: ruleConfig.priority,
        direction: ruleConfig.direction,
        access: ruleConfig.access,
        protocol: ruleConfig.protocol,
        sourcePortRange: ruleConfig.sourcePortRange,
        destinationPortRange: ruleConfig.destinationPortRange,
        sourceAddressPrefix: ruleConfig.sourceAddressPrefix,
        destinationAddressPrefix: ruleConfig.destinationAddressPrefix,
      });
    }

    this.id = nsg.id;
    this.name = nsg.name;
  }

  // Method to associate the network security group to a subnet
  public associateToSubnet(subnet: Subnet) {
    new SecurityGroupAssociations(this, subnet.name, {
      subnetId: subnet.id,
      networkSecurityGroupId: this.id,
    });
  }

  // Method to associate the network security group to a network interface
  public associateToNetworkInterface(networkInterface: NetworkInterface) {
    new SecurityGroupAssociations(this, networkInterface.name, {
      networkInterfaceId: networkInterface.id,
      networkSecurityGroupId: this.id,
    });
  }
}

/**
 * Properties for associating Azure Network Security Groups with subnets and network interfaces.
 */
export interface SecurityGroupAssociationsProps {
  /**
   * The ID of the network security group to be associated.
   */
  readonly networkSecurityGroupId: string;

  /**
   * Optional subnet ID to associate with the network security group.
   * If provided, the security group will be associated with this subnet.
   */
  readonly subnetId?: string;

  /**
   * Optional network interface ID to associate with the network security group.
   * If provided, the security group will be associated with this network interface.
   */
  readonly networkInterfaceId?: string;
}

export class SecurityGroupAssociations extends Construct {
  constructor(
    scope: Construct,
    id: string,
    props: SecurityGroupAssociationsProps,
  ) {
    super(scope, id);
    // If subnetId is provided, create a SubnetNetworkSecurityGroupAssociation
    if (props.subnetId) {
      new SubnetNetworkSecurityGroupAssociation(this, "subassociation", {
        subnetId: props.subnetId,
        networkSecurityGroupId: props.networkSecurityGroupId,
      });
    }

    // If networkInterfaceId is provided, create a NetworkInterfaceSecurityGroupAssociation
    if (props.networkInterfaceId) {
      new NetworkInterfaceSecurityGroupAssociation(this, "nicassociation", {
        networkInterfaceId: props.networkInterfaceId,
        networkSecurityGroupId: props.networkSecurityGroupId,
      });
    }
  }
}
