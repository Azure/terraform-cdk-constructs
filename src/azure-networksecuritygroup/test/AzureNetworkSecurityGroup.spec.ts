import { Testing, TerraformStack} from 'cdktf';
import 'cdktf/lib/testing/adapters/jest';
import { AzureNetworkSecurityGroup } from '..';
import {AzurermProvider} from "@cdktf/provider-azurerm/lib/provider";
import { exampleAzureNetworkSecurityGroup } from './ExampleAzureNetworkSecurityGroup';
import {PreconfiguredRules} from "../preconfiguredRules";

describe('Azure Network Security Group With Defaults', () => {
  let stack: TerraformStack;
  let fullSynthResult: any;

  beforeEach(() => {
    const app = Testing.app();
    stack = new TerraformStack(app, "test");

    new AzurermProvider(stack, "azureFeature", {features: {}});

    // Create a network security group with the defined rules
    new AzureNetworkSecurityGroup(stack, 'testAzureNetworkSecurityGroupDefaults', {
      name: 'my-nsg',
      location: "eastus",
      resourceGroupName: "rgtest",
      rules: [PreconfiguredRules.SSH],
    });

    fullSynthResult = Testing.fullSynth(stack); // Save the result for reuse
  });

  it("renders an Azure Network Security Group with defaults and checks snapshot", () => {
    expect(
      Testing.synth(stack)
    ).toMatchSnapshot(); // Compare the already prepared stack
  });

  it("check if the produced terraform configuration is valid", () => {
    expect(fullSynthResult).toBeValidTerraform(); // Use the saved result
  });

  it("check if this can be planned", () => {
    expect(fullSynthResult).toPlanSuccessfully(); // Use the saved result
  });
});


describe('Azure Network Security Group Example', () => {
  
  it("renders the Example Network Security Group and checks snapshot", () => {
  
    expect(
      Testing.synth(new exampleAzureNetworkSecurityGroup(Testing.app(), "testNetworkSecurityGroupExample"))
      
    ).toMatchSnapshot();
});

  it("check if the produced terraform configuration is valid", () => {
    // We need to do a full synth to plan the terraform configuration
    expect(Testing.fullSynth(new exampleAzureNetworkSecurityGroup(Testing.app(), "testNetworkSecurityGroupExample"))).toBeValidTerraform();
  });

  it("check if this can be planned", () => {
  
    
    // We need to do a full synth to plan the terraform configuration
    expect(Testing.fullSynth(new exampleAzureNetworkSecurityGroup(Testing.app(), "testNetworkSecurityGroupExample"))).toPlanSuccessfully();
  });
});