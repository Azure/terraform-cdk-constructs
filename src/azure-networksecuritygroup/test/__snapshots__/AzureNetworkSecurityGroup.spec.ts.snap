// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Azure Network Security Group Example renders the Example Network Security Group and checks snapshot 1`] = `
"{
  "output": {
    "nsg_name": {
      "value": "\${azurerm_network_security_group.nsg_my-nsg_3DA143E1.name}"
    },
    "resource_group_name": {
      "value": "\${azurerm_resource_group.rg.name}"
    },
    "ssh_rule_name": {
      "value": "SSH"
    }
  },
  "provider": {
    "azurerm": [
      {
        "features": {
        }
      }
    ]
  },
  "resource": {
    "azurerm_network_security_group": {
      "nsg_my-nsg_3DA143E1": {
        "location": "eastus",
        "name": "my-nsg",
        "resource_group_name": "\${azurerm_resource_group.rg.name}"
      }
    },
    "azurerm_network_security_rule": {
      "nsg_RDP_802FDDEB": {
        "access": "Allow",
        "destination_address_prefix": "*",
        "destination_port_range": "3389",
        "direction": "Inbound",
        "name": "RDP",
        "network_security_group_name": "\${azurerm_network_security_group.nsg_my-nsg_3DA143E1.name}",
        "priority": 505,
        "protocol": "Tcp",
        "resource_group_name": "\${azurerm_resource_group.rg.name}",
        "source_address_prefix": "10.0.0.0/24",
        "source_port_range": "*"
      },
      "nsg_SSH_56EE6BE4": {
        "access": "Allow",
        "destination_address_prefix": "VirtualNetwork",
        "destination_port_range": "22",
        "direction": "Inbound",
        "name": "SSH",
        "network_security_group_name": "\${azurerm_network_security_group.nsg_my-nsg_3DA143E1.name}",
        "priority": 1001,
        "protocol": "Tcp",
        "resource_group_name": "\${azurerm_resource_group.rg.name}",
        "source_address_prefix": "10.23.15.38",
        "source_port_range": "*"
      }
    },
    "azurerm_resource_group": {
      "rg": {
        "location": "eastus",
        "name": "rg-test"
      }
    },
    "azurerm_subnet": {
      "subnet1": {
        "address_prefixes": [
          "10.0.1.0/24"
        ],
        "name": "subnet1",
        "resource_group_name": "\${azurerm_resource_group.rg.name}",
        "virtual_network_name": "\${azurerm_virtual_network.vnet.name}"
      },
      "subnet2": {
        "address_prefixes": [
          "10.0.2.0/24"
        ],
        "name": "subnet2",
        "resource_group_name": "\${azurerm_resource_group.rg.name}",
        "virtual_network_name": "\${azurerm_virtual_network.vnet.name}"
      }
    },
    "azurerm_subnet_network_security_group_association": {
      "nsg_TfTokenTOKEN36_subassociation_1654316F": {
        "network_security_group_id": "\${azurerm_network_security_group.nsg_my-nsg_3DA143E1.id}",
        "subnet_id": "\${azurerm_subnet.subnet1.id}"
      },
      "nsg_TfTokenTOKEN38_subassociation_A605BE6F": {
        "network_security_group_id": "\${azurerm_network_security_group.nsg_my-nsg_3DA143E1.id}",
        "subnet_id": "\${azurerm_subnet.subnet2.id}"
      }
    },
    "azurerm_virtual_network": {
      "vnet": {
        "address_space": [
          "10.0.0.0/16"
        ],
        "location": "\${azurerm_resource_group.rg.location}",
        "name": "vnet-test",
        "resource_group_name": "\${azurerm_resource_group.rg.name}"
      }
    }
  },
  "terraform": {
    "required_providers": {
      "azurerm": {
        "source": "azurerm",
        "version": "3.66.0"
      }
    }
  }
}"
`;

exports[`Azure Network Security Group With Defaults renders an Azure Network Security Group with defaults and checks snapshot 1`] = `
"{
  "provider": {
    "azurerm": [
      {
        "features": {
        }
      }
    ]
  },
  "resource": {
    "azurerm_network_security_group": {
      "testAzureNetworkSecurityGroupDefaults_my-nsg_C4572744": {
        "location": "eastus",
        "name": "my-nsg",
        "resource_group_name": "rgtest"
      }
    },
    "azurerm_network_security_rule": {
      "testAzureNetworkSecurityGroupDefaults_SSH_B74D5EF3": {
        "access": "Allow",
        "destination_address_prefix": "*",
        "destination_port_range": "22",
        "direction": "Inbound",
        "name": "SSH",
        "network_security_group_name": "\${azurerm_network_security_group.testAzureNetworkSecurityGroupDefaults_my-nsg_C4572744.name}",
        "priority": 501,
        "protocol": "Tcp",
        "resource_group_name": "rgtest",
        "source_address_prefix": "*",
        "source_port_range": "*"
      }
    }
  },
  "terraform": {
    "required_providers": {
      "azurerm": {
        "source": "azurerm",
        "version": "3.66.0"
      }
    }
  }
}"
`;
